mqtt/mqtt-bridge/src/persist/waking_state/memory.rs

```diff
@@ -0,0 +1,56 @@
+use std::{
+    cmp::min,
+    collections::{HashMap, VecDeque},
+    task::Waker,
+};
+
+use mqtt3::proto::Publication;
+
+use crate::persist::{waking_state::StreamWakeableState, Key, PersistError};
+
+/// When elements are retrieved they are moved to the in flight collection.
+pub struct WakingMemoryStore {
+    queue: VecDeque<(Key, Publication)>,
+    in_flight: HashMap<Key, Publication>,
+    waker: Option<Waker>,
+}
+
+impl WakingMemoryStore {
+    pub fn new() -> Self {
+        WakingMemoryStore {
+            queue: VecDeque::new(),
+            in_flight: HashMap::new(),
+            waker: None,
+        }
+    }
+}
+
+impl StreamWakeableState for WakingMemoryStore {
+    fn insert(&mut self, key: Key, value: Publication) -> Result<(), PersistError> {
+        self.queue.push_back((key, value));
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+
+    fn batch(&mut self, count: usize) -> Result<VecDeque<(Key, Publication)>, PersistError> {
+        let count = min(count, self.queue.len());
+        let output: VecDeque<_> = self.queue.drain(..count).collect();
+        self.in_flight.extend(output.clone().into_iter());
```

Comment:
nit: this is a hot path. Would it be better to iterate over Drain and insert into both output and in_flight instead of extra cloning of VecDeque?

Reply:
You are saying this is better because rather than iterating over drain twice like we are doing here, we only have to iterate once? If there is a sneaky optimization with the clone() function, this new approach might be less performant. I'm not sure so I'll do whatever you think it is most likely best.

Reply:
we can keep it as is for now, and maybe play with it when we have bench tests.

Reply:
I also not sure what @vadim-kovalyov mentioned because we still need a copy in both places: current batch to iterate and inflight messages to avoid dups.

Reply:
It's probably better describe with the code :)
```rust
    fn batch(&mut self, count: usize) -> Result<VecDeque<(Key, Publication)>, PersistError> {
        let count = min(count, self.queue.len());
        let mut output = VecDeque::with_capacity(count);
        for (key, publication) in self.queue.drain(..count) {
            self.loaded.insert(key, publication.clone());
            output.push_back((key, publication));
        }

        Ok(output)
    }
```

Reply:
but also, why do we need to keep the publications in `self.loaded`? isn't just keys enough?

Reply:
It is there to facilitate returning the publication when ```remove() -> Publication``` is called. It is only absolutely necessary for the in-memory case. We could simplify it to just keys for the disk case though.

Thinking more about how we are going to use the persistor, I think the ```remove()``` method can return void. I'm going to make this change. This means we will only need the keys like you are suggesting.